### ISSUE TITLE
Non-Github account creation

### ISSUE THREAD

**Issue by alexander255** [rust-lang/crates.io#326]
http://doc.crates.io/crates-io.html says:

    Acquiring an API token
    
    First thing’s first, you’ll need an account on crates.io to acquire an API token. To do so, visit the home page and log in via a GitHub account (required for now).

Any plans to change this? Not publishing on crates.io meanwhile…

**Comment by steveklabnik** [comment]
I don't know of anyone working on additional logins.

Given that you already have a github account, what specifically is the issue? I can think of a few, I'm interested in yours, specifically.

**Comment by alexander255** [comment]
Particularly my problem is that I don't have crates.io account and that I don't think GitHub has, or should have, anything to do with *publishing Rust crates*. How is there any (not artificially induced) overlap between what crates.io does and what GitHub does?
There is practicality none, except for the fact that crates.io developers seem to really have taken a liking in using GitHub for everything and one thing more. Using GitHub as [public git file storage](https://github.com/rust-lang/crates.io-index) is similar to this, but in a much less worrying way: GitHub remains a strictly swappable component. Don't like GitHub anymore one day? Just adapt @bors to something else and within a day or two you're done migrating. Also that's something that can easily done by a core developer; switching the user backend? Not so much…

@steveklabnik Is there any benefit to using GitHub Account API, aside from the fact that it happened to be the first user database that attracted some developer's eye?

**Comment by steveklabnik** [comment]
> except for the fact that crates.io developers seem to really have taken a liking in using GitHub for everything and one thing more

The only things dealing with github are the index, which is a plain git repository, and therefore easy to swap, and the login system, which is the usual oauth based thing.

The reasons for choosing GitHub for logins was the same as any site that enables OAuth: new users don't have to trust us with storing passwords, nor making Yet Another Account on another website. There's no inherent reason that it has to stay GitHub only, other than nobody has bothered to actually implement it.

**Comment by alexander255** [comment]
… except that crates.io really isn't a lot more than an index with per-login management.
But we're actually not disagreeing on anything, so let's not argue over what each of us considers *a lot* in this context.

Since you seem to not be not very fond of the idea of "normal" logins (correct me if I'm wrong). Have you considered an OpenID login option instead?

**Comment by steveklabnik** [comment]
Well, there's also the hosting of all of the crates' source code, permissions around who and what is exactly allowed to publish, validating that Cargo.toml is filled out correctly......

I'm not opposed to a 'normal' login. I wouldn't be particularly pro-OpenID though, personally. But it's @alexcrichton 's opinion that matters most.

**Comment by alexcrichton** [comment]
Yes there's no particular reason that we don't have anything other than GitHub yet beyond that no one's actually implemented it. It was always the intent to have a variety of login options, and then you could link multiple login varieties to the same account (e.g. you can log in via oauth from either Twitter or GitHub)

**Comment by alexander255** [comment]
@alexcrichton: So if I were to come along with an OpenID Auth implementation you'd probably merge it (aside from the usual *getting-to-know-the-codebase*-mistakes of course)?
Could you give me some pointers to were it'd be a good idea to implement this in the current codebase?

**Comment by alexcrichton** [comment]
@alexander255 certainly! Github authorization is currently handled [around here](https://github.com/rust-lang/crates.io/blob/9f0516cc83d3e634651325b2158580bb656f7f0e/src/user/mod.rs#L145-L248), and I suspect that something like OpenID would be similar and we'd just want a landing page to pick which one you want ahead of time (instead of always going to github first)

**Comment by grawlinson** [comment]
So what's actually required to implement alternative logins? Just hooking into Gitlab's OAuth provider?

Would anyone be willing to mentor me if I choose to work on this?

**Comment by jtgeibel** [comment]
@grawlinson here are some of the places we have code that interacts with GitHub:

* `src/app.rs` - configuration of the OAuth client
* `src/controllers/user/session.rs` - endpoints so that the frontend can obtain the authorization URL to open in a popup, and to verify the code provided by the completed authorization
* `src/github.rs` - interaction with the GitHub API, mainly for determining team membership
* `src/models/team.rs` - If the login contains a ":" it is a GitHub team.  (A team member can publish and yank owned crates, but cannot add/remove other owners.)
* `src/models/owner.rs` - If the owner is not a team, then they have full access to the crate.
* `src/models/user.rs` - The user model includes fields such as: `gh_id`, `gh_login`, `gh_avatar`, `gh_access_token`

There are several tricky things to keep in mind.  It's been a while since I've looked at the details, but there is logic to handle renamed users and teams.  If I recall correctly, we use the `gh_id` field for users and update the `gh_login` if the user has renamed themselves on GitHub.  Teams are handled differently, so that if a team is deleted and a new one with the same name is added, then we pick up the new ID number for the team.  I expect that there would be similar issues when dealing with GitLab users and groups and that we would want to extract this somehow to be as generic as possible.

In addition to the backend work, there would also need to be a frontend interface to select the appropriate login provider.

Since this looks like a fairly large amount of work, I think we would want to split this up into multiple PRs, incrementally adding functionality and enabling the frontend interface in the last one.  A good place to start might be to see if there is a reasonable way to extract the GitHub specific fields from the user/owner/team models to abstract over multiple OAuth providers.

**Comment by grawlinson** [comment]
OK, I'll have a look at the models and get started on a PR.

**Comment by maghoff** [comment]
@grawlinson, did you get anywhere with this? :) Now that login (again?) is required for publishing, I think there's renewed interest in this. From the comment history it looks like you are working on it, but it's been a while since the last update. Can you confirm if you're still on the case? :)

**Comment by grawlinson** [comment]
I'm still interested in implementing this, but real life takes up all my free time at the moment.

I can probably spare a day or two per week, if anyone's willing to mentor/guide.

**Comment by tarcieri** [comment]
This really feels to me like something that should go through the RFC process. Identity and its relationship to access control are touchy topics (and also an area of personal expertise for me), and ones which should, IMO, get proper design and security review before implementation work starts.

**Comment by paulocsanz** [comment]
I would like to give another viewpoint about why we need more than github.

The company I work on uses github and needs you to setup 2fa on your account to access its repositories. I enabled it. My cellphone broke and I lost access to the account and the recovery code wasn't working.

I would have lost access to my crates and they would me unmantained forever without a warning to the user, this is a huge problem.

Thankfully I had access to the SSH keys used in my accoutn and was able to disable 2fa, but if I hadn't those crates would be unaccessible.

**Comment by grawlinson** [comment]
Agreed. I’ll start drafting and submit a RFC over at the rust-lang/rfcs repository.

**Comment by jolhoeft** [comment]
Last month I added OAuth support for several providers to my nickel.rs based site (https://virtuarasa.com/), so a lot of these issues are fresh in my mind, although I've not yet looked at the crates.io code yet. @jtgeibel 's outline of the issues above maps to much of what I needed to deal with. If it won't step on anyone's toes, I'd like to take a swing at this. My oauth identity module may generalize for this application (if so I'll make a separate crate of it).

I would treat linking accounts, e.g. having a login with Google the same as a login with Github, as a completely separate issue after multiple providers are supported.

Is there a desire to support local accounts, i.e. users create a username and password on crates.io?

**Comment by carols10cents** [comment]
> Is there a desire to support local accounts, i.e. users create a username and password on crates.io?

Not at this time, no.

**Comment by tarcieri** [comment]
@jolhoeft perhaps you could help @grawlinson work on an RFC?

**Comment by jolhoeft** [comment]
I've got some time to focus on this. Should I just draft an RFC and create a PR, or would it be better to post it here for discussion first? Also, is there a good example of a crates.io RFC to look at? There seem to be mostly language RFCs, which seem to be looking for somewhat different things than this would.

**Comment by grawlinson** [comment]
Looking at the RFC process, it may be more desirable to discuss it some more on this issue. Plus, this issue does have all the related links/issues/etc posted on it so far.

Would you say that "support for alternative OAuth providers" is a good starting point, @jolhoeft?

**Comment by jolhoeft** [comment]
Yes, maybe "Support Configurable OAuth providers". Sadly, login with OAuth is not particularly standardized, with most providers rolling their own identity schema. There is OpenIDConnect, which Google and I think PayPal support, but most others supply some custom defined Json blob. That said, I think we could manage to cover most providers with just configuration values, so providers could be enabled and disabled at startup or even while running.

Some question I need to go through the code to find out:

* What information are we getting from Github? I just needed the provider's user id, email address, and display name, which have always been available with a single api call.
  * Annoyingly, Google recently stopped providing the display name
* What struct are we using for users and where are we storing it, presumably a database table?
  * Can we extend the table for the OAuth, or is it better to add another table? I suspect the latter.
* (Not the code) What providers are important to support?
  * I initially supported five, and I fear I may have opened myself up for some operational headache, as they periodically change or break. Fewer may be less burden.

We are using diesel for persistence, conduit as a web framework, and a rather old version of oauth2. I may create a PR to upgrade oauth2 before I get too deep into this.

**Comment by jolhoeft** [comment]
BTW, how does crates.io interact with the private cargo repositories? That seems like that would impact this.

**Comment by tarcieri** [comment]
> Also, is there a good example of a crates.io RFC to look at? 

@jolhoeft one example is the RFC for signing registry index commits: https://github.com/rust-lang/rfcs/pull/2474

**Comment by yocorn** [comment]
Waiting for **GitLab** login or **email** login!

**Comment by yocorn** [comment]
Any Updates on this issue?

**Comment by jolhoeft** [comment]
I need a week free to code this up, but that doesn't look likely before late January. I stalled out on the RFC initially. I should probably have just create a pull request and let the discussion go from there.

**Comment by flowed** [comment]
# **Looking forward for GitLab auth!**

**Comment by recapitalverb** [comment]
It has been about 4 years and there's not much people working on this, and Microsoft's acquisition of GitHub makes it worse. Anyone?

**Comment by carols10cents** [comment]
@recapitalverb nope, this issue is still up for grabs if you'd like to work on it!

**Comment by Xaeroxe** [comment]
> This really feels to me like something that should go through the RFC process.

Has anyone written an RFC yet or is that step still needed?

**Comment by diceride** [comment]
Does it not bother anyone, that all crates and their related accounts are currently managed by _Microsoft_? You can't make this up. sus

**Comment by diceride** [comment]
Does it not bother anyone, that all crates and their related accounts are currently managed by Microsoft?

**Comment by realtimetodie** [comment]
Does it not bother anyone, that all crates and their related accounts are currently managed by _Microsoft_?

**Comment by AaronM04** [comment]
I would be happy to beta test such a solution when it's dev complete!

**Comment by ghost** [comment]
Will there be other ways to sign up?

I use GitLab and want access without needing to mirror my repo to GitHub or make it public.

**Comment by trevyn** [comment]
> Will there be other ways to sign up?
> 
> I use GitLab and want access without needing to mirror my repo to GitHub or make it public.

You need a GitHub account to get a publishing token, but it’s not otherwise connected; you don’t need to push any code to GitHub at all.

**Comment by carols10cents** [comment]
> Will there be other ways to sign up?

There will be other ways to sign up when someone has time to implement it.

**Comment by ghost** [comment]
> > Will there be other ways to sign up?
> > 
> > I use GitLab and want access without needing to mirror my repo to GitHub or make it public.
> 
> You need a GitHub account to get a publishing token, but it’s not otherwise connected; you don’t need to push any code to GitHub at all.

I want to push the crates with org(GitHub) and group name (GitLab). For example: `acme/project`. How to achieve that?

**Comment by WhyNotHugo** [comment]
@prnam That's unrelated to this topic; see https://github.com/rust-lang/crates.io/issues/8292.

**Comment by fluffy** [comment]
Pretty outrageous this has been open since 2016 and has not been fixed and you have to grant access to crates.io to read private project boards to be able to logon.

**Comment by WhyNotHugo** [comment]
@carols10cents Are you volunteering?

**Comment by fluffy** [comment]
So for slight context I am one of the CTO at Cisco and I do fund some open source work both inside and outside Cisco.  What might you suggest I could do if I wanted to get it so that it at least only asked for public data from github?

**Comment by fluffy** [comment]
I know nothing about the code for crates.io but at casual glance at code looks like changing the line in src/controllers/user/session.rs that reads 
```
.add_scope(Scope::new("read:org".to_string()))" 
```
and changing the scope from "read:org" to "user:email" or "read:user" might work. I could not see where an org scope was needed, but of course I don't understand the code. 

Would that be enough to solve this privacy issue ?

**Comment by Turbo87** [comment]
> I could not see where an org scope was needed, but of course I don't understand the code.

we currently allow GitHub org teams to be owners of crates and to figure out if a publishing used is a member of a team we need the `read:org` scope.

> Would that be enough to solve this privacy issue ?

you can find all the data we request from GitHub at https://github.com/rust-lang/crates.io/blob/916fb9213d7849bd15a6afd23fd0129c00b96ecf/crates/crates_io_github/src/lib.rs. if there anything in there that you consider too secret to share with crates.io?

**Comment by fluffy** [comment]
@Turbo87 - thank you kindly for pointer. I'll spend some time looking at that.

**Comment by WhyNotHugo** [comment]
[GitHub's documentation indicates that users are able to allow only certain scopes](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps#requested-scopes-and-granted-scopes):

> It's important to handle error cases where a user chooses to grant you less access than you originally requested. For example, applications can warn or otherwise communicate with their users that they will see reduced functionality or be unable to perform some actions.

However, I don't see this posibility in the authorization pop-up.

**Comment by programmerjake** [comment]
maybe have it configurable which scopes are requested? so e.g. you can select just `user:email` if you don't want to use any teams, and reauthorize with github later if you change your mind.

**Comment by carols10cents** [comment]
There already is a more specific issue for logging in without org read: https://github.com/rust-lang/crates.io/issues/3027

**Comment by ghost** [comment]
As a data point: I am leaving GitHub permanently and it is unclear what deleting my GitHub account will do to my crates.io account and package.

**Comment by Turbo87** [comment]
> it is unclear what deleting my GitHub account will do to my crates.io account and package.

it will make your account inaccessible and your package unmaintained

**Comment by inferiorhumanorgans** [comment]
> Yes there's no particular reason that we don't have anything other than GitHub yet beyond that no one's actually implemented it. It was always the intent to have a variety of login options, and then you could link multiple login varieties to the same account (e.g. you can log in via oauth from either Twitter or GitHub)

@alexcrichton We're coming up on a decade since your comment.  Is there an ETA to removing the hard dependency on GitHub?

**Comment by ofluffydev** [comment]
I was disappointed to see that crates.io still only supports GitHub login, and even more surprised to see that this has been an open issue for around ten years.

With the growing concerns around GitHub, especially regarding AI-related decisions and platform lock-in, I think a lot of people are more interested than ever in having a real path away from GitHub. I’m one of them.

It looks like there has been some recent progress here, but after reading through the discussion and PRs, I’m honestly a bit confused about the current state of the work. From what I can tell, one of the recent attempts may have gone off-track by reworking Docker-related pieces that are not actually part of the production setup. I saw comments indicating that Docker is not used in production for crates.io, while the PR seemed to be targeting or restructuring a production Docker image that does not appear to exist.

No offense is intended toward anyone who has worked on this. I know this is a non-trivial change, and I appreciate that people have been trying to move it forward. I’m just trying to understand what still needs to be done and where the previous approach became blocked.

I’m planning to dig into this more this evening and try to untangle the current state of things as best I can. I strongly believe this change needs to happen, and I’m willing to help get it across the finish line.

:3

**Comment by Turbo87** [comment]
> I’m just trying to understand what still needs to be done

the current step in the process is that https://github.com/rust-lang/rfcs/pull/3946 needs to finish the final comment period and then we can start working on implementing that. the open PRs that you are referring to jumped several steps ahead and simply aren't mergeable at the moment. I recommend reading the RFC to understand why this is not a trivial process.

**Comment by carols10cents** [comment]
I am providing updates on the refactoring work that needs to be done here https://github.com/rust-lang/crates.io/issues/10611

**Comment by carols10cents** [comment]
https://github.com/rust-lang/rfcs/pull/3946 has been merged! I've started a tracking issue here: https://github.com/rust-lang/crates.io/issues/13758 That issue now has some sub-issues that I'd love people's help on.

**Comment by hwittenborn** [comment]
@carols10cents if you're able to get those sub-issues out, I'd love to help out. My team has a strong need for some other sign-in options :)

**Comment by martinetd** [comment]
See https://github.com/rust-lang/crates.io/issues/13758 for sub-issues (from the edited top comment, it's updated regularily)
